x-pack/plugins/security_solution/public/assistant/get_comments/stream/message_text.tsx

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import {
+  EuiMarkdownFormat,
+  EuiSpacer,
+  EuiText,
+  getDefaultEuiMarkdownParsingPlugins,
+  getDefaultEuiMarkdownProcessingPlugins,
+} from '@elastic/eui';
+import { css } from '@emotion/css';
+import classNames from 'classnames';
+import type { Code, InlineCode, Parent, Text } from 'mdast';
+import React from 'react';
+import type { Node } from 'unist';
+import { customCodeBlockLanguagePlugin } from '../custom_codeblock/custom_codeblock_markdown_plugin';
+import { CustomCodeBlock } from '../custom_codeblock/custom_code_block';
+
+interface Props {
+  content: string;
+  loading: boolean;
+}
+
+const ANIMATION_TIME = 1;
+
+const cursorCss = css`
+  @keyframes blink {
+    0% {
+      opacity: 0;
+    }
+    50% {
+      opacity: 1;
+    }
+    100% {
+      opacity: 0;
+    }
+  }
+
+  animation: blink ${ANIMATION_TIME}s infinite;
+  width: 10px;
+  height: 16px;
+  vertical-align: middle;
+  display: inline-block;
+  background: rgba(0, 0, 0, 0.25);
+`;
+
+const Cursor = () => (
+  <span data-test-subj="cursor" key="cursor" className={classNames(cursorCss, 'cursor')} />
+);
+
+// a weird combination of different whitespace chars to make sure it stays
+// invisible even when we cannot properly parse the text while still being
+// unique
+const CURSOR = ` ᠎  `;
+
+const loadingCursorPlugin = () => {
+  const visitor = (node: Node, parent?: Parent) => {
+    if ('children' in node) {
+      const nodeAsParent = node as Parent;
+      nodeAsParent.children.forEach((child) => {
+        visitor(child, nodeAsParent);
+      });
+    }
+
+    if (node.type !== 'text' && node.type !== 'inlineCode' && node.type !== 'code') {
+      return;
+    }
+
+    const textNode = node as Text | InlineCode | Code;
+
+    const indexOfCursor = textNode.value.indexOf(CURSOR);
+    if (indexOfCursor === -1) {
+      return;
+    }
+
+    textNode.value = textNode.value.replace(CURSOR, '');
+
+    const indexOfNode = parent?.children.indexOf(textNode) ?? 0;
+    parent?.children.splice(indexOfNode + 1, 0, {
+      type: 'cursor' as Text['type'],
+      value: CURSOR,
+    });
+  };
+
+  return (tree: Node) => {
+    visitor(tree);
+  };
+};
+
+const getPluginDependencies = () => {
+  const parsingPlugins = getDefaultEuiMarkdownParsingPlugins();
+
+  const processingPlugins = getDefaultEuiMarkdownProcessingPlugins();
+
+  const { components } = processingPlugins[1][1];
+
+  processingPlugins[1][1].components = {
+    ...components,
+    cursor: Cursor,
+    customCodeBlock: (props) => {
+      return (
+        <>
+          <CustomCodeBlock value={props.value} />
+          <EuiSpacer size="m" />
+        </>
+      );
+    },
```

Comment:
Looks like everything around the markdown parsing/processing got moved over correctly, however I'm not seeing the codeblock actions show up anymore:

<p align="center">
  <img width="500" src="https://github.com/elastic/kibana/assets/2946766/f916b37e-940e-45a6-8829-9add510b4ec2" />
</p> 

I confirmed the custom codeblock is rendering as expected, so it is probably something downstream w/ `augmentMessageCodeBlocks`

https://github.com/elastic/kibana/blob/480ef875b88bc03c997206381a1896f9c17e28bb/x-pack/plugins/security_solution/public/assistant/helpers.tsx#L80-L97

You had mentioned some changes around `getMessageContentWithReplacements()`, which this ends up calling -- perhaps it's coming from there?

Reply:
Ok I've got this sort of figured out. First off, I dropped the className that the `getControlContainer` was looking for, `className={'message-${index}'}`. I [put the class in place](https://github.com/elastic/kibana/pull/169729/files#diff-1f71044c1996550b210d56e0a5da05a496845c36523c732b8de83517b3df4091R167-R168), and it is working again for NON-streaming.

Putting a `debugger` in the `getControlContainer` while streaming is turned on, when our comments changes from `reader` to `content` there seems to be a flash where the `.euiCodeBlock__controls` disappears. As soon as the debugger is played, the controls appear again. ❓ 

Playing with this, I tried switching from `useLayoutEffect` to `useEffect`, but the same issue happened. I added a `setTimeout`, and experimented with the timeout length and even a timeout of 0 seems to resolve the issue. Kind of a render bandaid that appears to be holding... what do you think? Any better ideas?

Reply:
Discussed offline w/ Steph -- this all looks good and even improves the inconsistencies with the current experience, so all the better! 🙂